### PR TITLE
Remove extensions/v1beta1 Job

### PR DESCRIPTION
**Pull request by soltysh**
Fixes https://github.com/kubernetes/kubernetes/issues/32763. This endpoint was deprecated in 1.5 and was planned to be removed in 1.6. 

**Release note**:
```release-note
Remove extensions/v1beta1 Jobs resource, and job/v1beta1 generator.
```

### CODE REVIEW

**Comment by soltysh** [comment]
@kubernetes/sig-apps @erictune ptal

**Comment by soltysh** [comment]
I'm tentatively assigning this to Eric, but he may choose somebody else from @kubernetes/sig-apps to be reviewer.

**Comment by kargakis** [comment]
@kubernetes/sig-api-machinery may also be interested

**Comment by soltysh** [comment]
Removed missed `job/v1beta1` generator test in `test-cmd.sh`.

**Comment by soltysh** [comment]
Fixed integration compilation error.

**Comment by soltysh** [comment]
Rebased.

**Comment by soltysh** [comment]
Rebased again, due to failures.

**Comment by soltysh** [comment]
Rebased again, due to failures.

**Comment by soltysh** [comment]
Rebased.

**Comment by soltysh** [comment]
@deads2k any ideas why this sometimes fails:
```
defaulting_test.go:153: expected rbac.authorization.k8s.io/v1alpha1, Kind=ClusterRoleBindingList to trigger defaulting due to fuzzing
```

**Comment by soltysh** [comment]
@k8s-bot unit test this

**Comment by kargakis** [comment]
> @deads2k any ideas why this sometimes fails:

It was the same for me in https://github.com/kubernetes/kubernetes/pull/37959. Not sure how but I rerun the generation scripts and it got fixed.

**Comment by soltysh** [comment]
@kargakis tell me about it, when I tried during the day it failed every single time, then I re-run it in the evening and it magically worked, without any changes, that's a bit worrisome.

**Comment by soltysh** [comment]
@k8s-bot unit test this

File: test/e2e/batch_v1_jobs.go [hunk 1]
```diff
@@ -1,342 +0,0 @@
-/*
```
Comment (caesarxuchao) [inline test/e2e/batch_v1_jobs.go]: Why is this test removed?

**Comment by caesarxuchao** [comment]
Is the commit "Remove extensions/v1beta1 Job - staging client" generated by running `hack/update-staging-client-go.sh`

**Comment by caesarxuchao** [comment]
One question regarding the removed test/e2e/batch_v1_jobs.go, otherwise LGTM.

@janetkuo could you double check the kubectl changes?

File: pkg/kubectl/cmd/run.go [hunk 2]
```diff
@@ -220,18 +220,13 @@ func Run(f cmdutil.Factory, cmdIn io.Reader, cmdOut, cmdErr io.Writer, cmd *cobr
 		case api.RestartPolicyOnFailure:
 			if contains(resourcesList, batchv1.SchemeGroupVersion.WithResource("jobs")) {
 				generatorName = "job/v1"
-			} else if contains(resourcesList, v1beta1.SchemeGroupVersion.WithResource("jobs")) {
```
Comment (janetkuo) [inline pkg/kubectl/cmd/run.go]: We'll need to update generator doc as well http://kubernetes.io/docs/user-guide/kubectl-conventions/#generators (in a separate PR and in the next release, of course)

File: pkg/kubectl/run.go [hunk 3]
```diff
@@ -193,27 +190,9 @@ func getEnvs(genericParams map[string]interface{}) ([]api.EnvVar, error) {
 	return envs, nil
 }
 
-func getV1Envs(genericParams map[string]interface{}) ([]v1.EnvVar, error) {
```
Comment (janetkuo) [inline pkg/kubectl/run.go]: Why remove `getV1Envs`?

File: pkg/kubectl/run.go [hunk 4]
```diff
@@ -451,13 +333,13 @@ func (CronJobV2Alpha1) Generate(genericParams map[string]interface{}) (runtime.O
...
 
-	podSpec, err := makeV1PodSpec(params, name)
+	podSpec, err := makePodSpec(params, name)
 	if err != nil {
 		return nil, err
 	}
 
-	imagePullPolicy := v1.PullPolicy(params["image-pull-policy"])
-	if err = updateV1PodContainers(params, args, envs, imagePullPolicy, podSpec); err != nil {
+	imagePullPolicy := api.PullPolicy(params["image-pull-policy"])
```
Comment (janetkuo) [inline pkg/kubectl/run.go]: We should use `v1.` instead of `api.`?

**Comment by janetkuo** [comment]
Haven't finished my review. @soltysh would you move the commits unrelated to "Remove extensions/v1beta1 Job" to a separate PR (for example, the ones about `CronJob`)? It'd make it easier to review changes

File: test/e2e/batch_v1_jobs.go (hunk 1 above)
Comment (soltysh) [inline test/e2e/batch_v1_jobs.go]: We have identical in `test/e2e/jobs.go`, originally when the two were written they were testing different endpoints (jobs.go - extensions/v1beta1 and batch_v1_jobs.go - batch/v1 Jobs), but from what I see, some time ago both were switched to Batch client and that's how it's left. I'm removing this one and leaving the other, to keep the naming scheme in-line with other resources.

**Comment by soltysh** [comment]
> Is the commit "Remove extensions/v1beta1 Job - staging client" generated by running hack/update-staging-client-go.sh?

@caesarxuchao yes

File: pkg/kubectl/cmd/run.go (hunk 2 above)
Comment (soltysh) [inline pkg/kubectl/cmd/run.go]: Correct, that's a followup to this one. I need to check the entire doc and make sure it's not using extensions/v1beta anywhere.

File: pkg/kubectl/run.go (hunk 3 above)
Comment (soltysh) [inline pkg/kubectl/run.go]: Back when Job internals were in extensions, and I was adding generator for `batch/v1` I had to create a Job targeted against that endpoint, which resulted in having copies of several methods in `run.go`. With this PR I'm removing those duplicates.

File: pkg/kubectl/run.go (hunk 4 above)
Comment (soltysh) [inline pkg/kubectl/run.go]: See my other answer in v1 vs regular api.

**Comment by soltysh** [comment]
> Haven't finished my review. @soltysh would you move the commits unrelated to "Remove extensions/v1beta1 Job" to a separate PR (for example, the ones about CronJob)? It'd make it easier to review changes

I have to rebase, so I guess it's not a problem to remove those and submit as followups.

**Comment by soltysh** [comment]
Rebased, but due to generation problems I had to include change to the generation (@caesarxuchao see first commit).

**Review by janetkuo — APPROVED** [review]
kubectl changes LGTM

File: test/e2e/batch_v1_jobs.go (hunk 1 above)
Comment (caesarxuchao) [inline test/e2e/batch_v1_jobs.go]: Ack.

**Comment by caesarxuchao** [comment]
@k8s-bot node e2e test this

**Pull request merged**